### PR TITLE
Handle placement priority

### PR DESCRIPTION
**Pull request by davidfischer**
Allow multiple placements to have a priority (`data-ea-priority`). When multiple prioritized placements are on the page, only one will be chosen by the server and returned taking into account the inventory and the priority placed by the client.

### CODE REVIEW

**Comment by ericholscher** [comment]
I honestly think Copilot did a better job reviewing this than I would have 🤷 It's hard for me to parse this kind of dense JS unfortunately, and it's using some syntax I'm not super familiar with.

File: docs/index.rst [hunk 1]
```diff
@@ -53,6 +53,11 @@ The following data attributes are supported on the ad placement element:
     A placement identifier. If you define an ``id`` and :ref:`enable placements reporting <placements>`,
     this will allow you to see reports for each ``id``.
 
+``data-ea-priority`` (optional)
+    A numerical priority for the placement. If multiple placements on a page define a priority,
+    the server will choose only one ad to return for the group based on the priority and available inventory.
+    Setting a priority when the page has a single placement will have no effect.
```
[the remark this answers is not in the document]
Comment (davidfischer) [reply docs/index.rst]: Added to the docs and validated.

File: index.js [hunk 2]
```diff
@@ -1029,39 +1123,61 @@ export function load_placements(force_load = false) {
...
+    ) {
+      if (!priority_groups[placement.publisher]) {
+        priority_groups[placement.publisher] = [];
       }
+      priority_groups[placement.publisher].push(placement);
+    }
+  });
 
+  Object.values(priority_groups).forEach((group) => {
+    if (group.length > 0) {
```
[the remark this answers is not in the document]
Comment (davidfischer) [reply index.js]: I changed this pretty considerably because I think handling multiple placements with different publishers on the same page shouldn't be supported.

File: index.js [hunk 3]
```diff
@@ -1029,39 +1123,61 @@ export function load_placements(force_load = false) {
...
+      placement &&
+      placement.priority !== null &&
+      (force_load || !placement.load_manually)
+    ) {
+      if (!priority_groups[placement.publisher]) {
+        priority_groups[placement.publisher] = [];
       }
+      priority_groups[placement.publisher].push(placement);
+    }
+  });
```
[the remark this answers is not in the document]
Comment (davidfischer) [reply index.js]: I changed this pretty considerably because I think handling multiple placements with different publishers on the same page shouldn't be supported.

File: index.js [hunk 4]
```diff
@@ -574,7 +582,9 @@ export class Placement {
     // Detect the keywords
     this.keywords = this.keywords.concat(this.detectKeywords());
 
-    return this.fetch()
+    let fetchPromise = this.fetchPromise || this.fetch();
```
[the remark this answers is not in the document]
Comment (davidfischer) [reply index.js]: This is OK I think. What this means is that only the winning ad position can rotate. It doesn't go into a priority again with all the different ad placements.

**Comment by davidfischer** [comment]
Overall, happy with how this turned out. I was able to test a number of edge cases successfully. There's still one I'd like to test before pushing this out but it requires a server-side fix (https://github.com/readthedocs/ethical-ad-server/pull/1143).

**Comment by davidfischer** [comment]
I verified this worked as intended after the server side fix was merged. I added docs. This is ready for review and release.
